website/blog/2025/05-05-arbiter-type-class/index.md

```diff
@@ -0,0 +1,280 @@
+This allows us to avoid passing around the type class instance explicitly.
+Instead, we can define the type class instance as an implicit value and the compiler will automatically find it for us.
+
+Let us rewrite our typeclass instances as implicit values:
+
+```scala
+implicit val arbiterClientArbitrable: Arbitrable[ArbiterClient] =
+  new Arbitrable[ArbiterClient] {
+    def request(a: ArbiterClient) = a.request
+    def grant(a: ArbiterClient, value: Bool) = a.grant := value
+  }
+
+// In chisel3.util, the type is DecoupledIO while we construct instances of it with Decoupled.
+// Note that this is a def because DecoupledIO itself takes a type parameter.
+implicit def decoupledArbitrable[T <: Data]: Arbitrable[DecoupledIO[T]] =
+  new Arbitrable[DecoupledIO[T]] {
+    def request(a: DecoupledIO[T]) = a.valid
+    def grant(a: DecoupledIO[T], value: Bool) = a.ready := value
+  }
+```
+
+And then we can refactor the arbiter to use the implicit type class instance:
+
+```scala
+class GenericPriorityArbiter[A <: Data](nClients: Int, clientType: A)(implicit arbitrable: Arbitrable[A]) extends Module {
+  val clients = IO(Vec(nClients, Flipped(clientType)))
+
+  val requests = clients.map(arbitrable.request(_))
+  val granted = PriorityEncoderOH(requests)
+
+  for ((client, grant) <- clients.zip(granted)) {
+    arbitrable.grant(client, grant)
+  }
+}
+```
+
+Now, we can instantiate the arbiter without passing the type class instance explicitly:
+
+```scala
+val clients1 = IO(Vec(4, Flipped(new ArbiterClient)))
+val arbiter1 = Module(new GenericPriorityArbiter(4, new ArbiterClient))
+arbiter1.clients :<>= clients1
+
+val clients2 = IO(Vec(4, Flipped(Decoupled(UInt(8.W)))))
+val arbiter2 = Module(new GenericPriorityArbiter(4, Decoupled(UInt(8.W))))
+arbiter2.clients :<>= clients2
+```
+
+This is much cleaner and more readable.
+
+Scala also has special syntax for implicit typeclass instances:
+
+```scala
+class GenericPriorityArbiter[A <: Data : Arbitrable](nClients: Int, clientType: A) extends Module {
+  ...
+}
+```
+
+This is equivalent to the previous definition, but is more concise.
+Note that unlike the version with the implicit argument, this one does not bind a variable name for the implicit argument.
+
+In the body of `GenericPriorityArbiter`, we can get a reference to the implicity value by calling `implicitly[Arbitrable[A]]`:
+```scala
+  val arbitrable = implicitly[Arbitrable[A]]
+```
+
+Note that Scala has rules for _implicit resolution_ for how to find the type class instance for a given type.
+As a general rule, you should define implicit type class instances in the companion object of the type they are for, or in the companion object for the type class itself.
+
+For example, since `DecoupledIO` is defined in Chisel itself, you could define the implicit value in the companion object for `Arbitrable`:
+```scala
+object Arbitrable {
+  implicit def decoupledArbitrable[T <: Data]: Arbitrable[DecoupledIO[T]] = ...
+}
+```
+
```

Comment:
Or if you are defining a new type you want to make arbitrable, you could include it in the package object for that type

Isn't this the more common case? Can we show it?